### PR TITLE
Add cuda_aware_mpi * cdt

### PR DESCRIPTION
**Pull request by jgphpc**
Fix for https://jira.cscs.ch/browse/VCMSA-190

### CODE REVIEW

File: checks/prgenv/cuda/cuda_aware_mpi.py [hunk 1]
```diff
@@ -138,3 +139,118 @@ def set_num_tasks(self):
         self.num_tasks_per_node = 1
         self.num_gpus_per_node = 1
         self.executable_opts = ['-t %d 1' % self.num_tasks]
+
+
+@rfm.simple_test
+class cuda_aware_mpi_check_cpe(rfm.RegressionTest):
```
Comment (jgphpc) [inline checks/prgenv/cuda/cuda_aware_mpi.py:145]:
```suggestion
class cuda_aware_mpi_check_cdt(rfm.RegressionTest):
```

File: checks/prgenv/cuda/cuda_aware_mpi.py [hunk 2]
```diff
@@ -138,3 +139,118 @@ def set_num_tasks(self):
...
+        'MPICH_VERSION_DISPLAY': '1',
+        'LD_LIBRARY_PATH': '$CRAY_LD_LIBRARY_PATH:$LD_LIBRARY_PATH',
+    }
+    prebuild_cmds = [
+        'rm -fr MATLAB* series c++11_cuda',
+        'cd posts/cuda-aware-mpi-example/src'
+    ]
+    build_system = 'Make'
+    maintainers = ['@ekouts', '@jgphpc']
+    tags = {'production', 'scs'}
```
Comment (jgphpc) [inline checks/prgenv/cuda/cuda_aware_mpi.py:165]:
```suggestion
    tags = {'craype'}
```

**Comment by jgphpc** [comment]
93315 tests on daint may be too much...

**Comment by jgphpc** [comment]
2 checks fail but not because of this PR:

```
[       OK ] ( 1/25) cuda_aware_mpi_check_xc %cdt_info=cdt/22.05 %nvhpc_info=nvhpc-nompi/22.2 %gcc_info=gcc/11.2.0 /bec2d3aa @daint:gpu+PrgEnv-gnu
[       OK ] ( 2/25) cuda_aware_mpi_check_xc %cdt_info=cdt/22.05 %nvhpc_info=nvhpc-nompi/21.9 %gcc_info=gcc/11.2.0 /7475bd84 @daint:gpu+PrgEnv-gnu
[       OK ] ( 3/25) cuda_aware_mpi_check_xc %cdt_info=cdt/22.05 %nvhpc_info=nvhpc-nompi/21.5 %gcc_info=gcc/10.3.0 /1295a309 @daint:gpu+PrgEnv-gnu
[       OK ] ( 4/25) cuda_aware_mpi_check_xc %cdt_info=cdt/22.05 %nvhpc_info=nvhpc-nompi/21.3 %gcc_info=gcc/10.3.0 /33dda222 @daint:gpu+PrgEnv-gnu
[       OK ] ( 5/25) cuda_aware_mpi_check_xc %cdt_info=cdt/22.02 %nvhpc_info=nvhpc-nompi/22.2 %gcc_info=gcc/11.2.0 /d9a9284a @daint:gpu+PrgEnv-gnu
[       OK ] ( 6/25) cuda_aware_mpi_check_xc %cdt_info=cdt/22.02 %nvhpc_info=nvhpc-nompi/21.9 %gcc_info=gcc/11.2.0 /d3e806a2 @daint:gpu+PrgEnv-gnu
[       OK ] ( 7/25) cuda_aware_mpi_check_xc %cdt_info=cdt/22.02 %nvhpc_info=nvhpc-nompi/21.5 %gcc_info=gcc/10.3.0 /f2593d82 @daint:gpu+PrgEnv-gnu
[       OK ] ( 8/25) cuda_aware_mpi_check_xc %cdt_info=cdt/22.02 %nvhpc_info=nvhpc-nompi/21.3 %gcc_info=gcc/10.3.0 /5434095a @daint:gpu+PrgEnv-gnu
[       OK ] ( 9/25) cuda_aware_mpi_check_xc %cdt_info=cdt/21.09 %nvhpc_info=nvhpc-nompi/22.2 %gcc_info=gcc/11.2.0 /b4d733c0 @daint:gpu+PrgEnv-gnu
[       OK ] (10/25) cuda_aware_mpi_check_xc %cdt_info=cdt/21.09 %nvhpc_info=nvhpc-nompi/21.9 %gcc_info=gcc/11.2.0 /68048730 @daint:gpu+PrgEnv-gnu
[       OK ] (11/25) cuda_aware_mpi_check_xc %cdt_info=cdt/21.09 %nvhpc_info=nvhpc-nompi/21.3 %gcc_info=gcc/10.3.0 /74a49c0c @daint:gpu+PrgEnv-gnu
[       OK ] (12/25) cuda_aware_mpi_check_xc %cdt_info=cdt/20.08 %nvhpc_info=nvhpc-nompi/22.2 %gcc_info=gcc/11.2.0 /98d3b999 @daint:gpu+PrgEnv-gnu
[       OK ] (13/25) cuda_aware_mpi_check_xc %cdt_info=cdt/21.09 %nvhpc_info=nvhpc-nompi/21.5 %gcc_info=gcc/10.3.0 /53eb02ad @daint:gpu+PrgEnv-gnu
[       OK ] (14/25) cuda_aware_mpi_check /7282dac5 @daint:gpu+PrgEnv-cray
[       OK ] (15/25) cuda_aware_mpi_check_xc %cdt_info=cdt/20.08 %nvhpc_info=nvhpc-nompi/21.9 %gcc_info=gcc/11.2.0 /6fda3258 @daint:gpu+PrgEnv-gnu
[       OK ] (16/25) cuda_aware_mpi_check /7282dac5 @daint:gpu+PrgEnv-gnu
[       OK ] (17/25) cuda_aware_mpi_check /7282dac5 @daint:gpu+PrgEnv-nvidia
[       OK ] (18/25) cuda_aware_mpi_check_xc %cdt_info=cdt/20.08 %nvhpc_info=nvhpc-nompi/21.5 %gcc_info=gcc/10.3.0 /18b7535f @daint:gpu+PrgEnv-gnu
[       OK ] (19/25) cuda_aware_mpi_check_xc %cdt_info=cdt/20.08 %nvhpc_info=nvhpc-nompi/21.3 %gcc_info=gcc/10.3.0 /34e965c7 @daint:gpu+PrgEnv-gnu
[       OK ] (20/25) cuda_aware_mpi_two_nodes_check /4c3007cd @daint:gpu+PrgEnv-cray
[       OK ] (23/25) cuda_aware_mpi_two_nodes_check /4c3007cd @daint:gpu+PrgEnv-gnu
[       OK ] (24/25) cuda_aware_mpi_one_node_check /c2502ea6 @daint:gpu+PrgEnv-cray
[       OK ] (25/25) cuda_aware_mpi_one_node_check /c2502ea6 @daint:gpu+PrgEnv-gnu

[     FAIL ] (21/25) cuda_aware_mpi_two_nodes_check /4c3007cd @daint:gpu+PrgEnv-nvidia
[     FAIL ] (22/25) cuda_aware_mpi_one_node_check /c2502ea6 @daint:gpu+PrgEnv-nvidia
```

File: checks/prgenv/cuda/cuda_aware_mpi.py [hunk 3]
```diff
@@ -59,11 +60,11 @@ def set_compilers(self):
             '-gencode arch=compute_{0},code=sm_{0}'.format(self.gpu_arch)
         )
         self.build_system.options = [
-            'CUDA_INSTALL_PATH=$CUDA_HOME',
-            'MPI_HOME=$CRAY_MPICH_PREFIX',
-            'GENCODE_FLAGS="%s"' % (gcd_flgs),
-            'MPICC="%s"' % self.current_environ.cc,
-            'MPILD="%s"' % self.current_environ.cxx
+            'CUDA_INSTALL_PATH=$CUDA_HOME',  # cuda_runtime.h
```
Comment (teojgo) [inline checks/prgenv/cuda/cuda_aware_mpi.py:63]:
```suggestion
            f'CUDA_INSTALL_PATH=$CUDA_HOME',  # cuda_runtime.h
```

File: checks/prgenv/cuda/cuda_aware_mpi.py [hunk 4]
```diff
@@ -59,11 +60,11 @@ def set_compilers(self):
...
+            'MPI_HOME=$CRAY_MPICH_PREFIX',  # mpi.h
```
Comment (teojgo) [inline checks/prgenv/cuda/cuda_aware_mpi.py:64]:
```suggestion
            f'MPI_HOME=$CRAY_MPICH_PREFIX',  # mpi.h
```

File: checks/prgenv/cuda/cuda_aware_mpi.py [hunk 5]
```diff
@@ -138,3 +139,129 @@ def set_num_tasks(self):
...
         self.executable_opts = ['-t %d 1' % self.num_tasks]
+
+
+def find_cdts(valid_systems, valid_prog_environs, modulename):
+    # TODO: use rt.runtime().system.partitions[] ?
+    modulefiles = []
+    modulefile_info = rfm_util.find_modules(modulename)
+    for modulefile in modulefile_info:
+        if modulefile[0] in valid_systems and modulefile[1] in \
+           valid_prog_environs:
```
Comment (teojgo) [inline checks/prgenv/cuda/cuda_aware_mpi.py:150]:
```suggestion
        if (modulefile[0] in valid_systems and
            modulefile[1] in valid_prog_environs):
```

File: checks/prgenv/cuda/cuda_aware_mpi.py [hunk 6]
```diff
@@ -138,3 +139,129 @@ def set_num_tasks(self):
...
+        ]
+
+    @run_before('compile')
+    def extract_versions(self):
+        cmd1 = "ldd ../bin/jacobi_cuda_aware_mpi | grep libcudart "
+        cmd2 = "awk '{print \"ls -l \",$3}'"
+        cmd3 = "sh"
+        cmd4 = "awk '{print $11}'"
+        self.rpt = os.path.join(self.stagedir, 'rpt')
+        cmd = "{} | {} | {} | {} > {}".format(cmd1, cmd2, cmd3, cmd4, self.rpt)
```
Comment (teojgo) [inline checks/prgenv/cuda/cuda_aware_mpi.py:235]:
```suggestion
        cmd = f'{cmd1} | {cmd2} | {cmd3} | {cm4} > {rpt}'
```

File: checks/prgenv/cuda/cuda_aware_mpi.py [hunk 7]
```diff
@@ -138,3 +139,129 @@ def set_num_tasks(self):
...
+            f'MPILD="{self.current_environ.cxx} {link_flags}"',
+        ]
+
+    @run_before('compile')
+    def extract_versions(self):
+        cmd1 = "ldd ../bin/jacobi_cuda_aware_mpi | grep libcudart "
+        cmd2 = "awk '{print \"ls -l \",$3}'"
+        cmd3 = "sh"
+        cmd4 = "awk '{print $11}'"
+        self.rpt = os.path.join(self.stagedir, 'rpt')
```
Comment (teojgo) [inline checks/prgenv/cuda/cuda_aware_mpi.py:234]:
```suggestion
        rpt = os.path.join(self.stagedir, 'rpt')
```

File: checks/prgenv/cuda/cuda_aware_mpi.py [hunk 8]
```diff
@@ -138,3 +139,129 @@ def set_num_tasks(self):
...
+        cuda_path = f'/opt/nvidia/hpc_sdk/Linux_x86_64/{nvhpc_version}/cuda'
+        link_flags = (
+            # add -lcuda to avoid segmentation fault
+            f'-Wl,-rpath={cuda_path}/lib64/ -lcuda '
+            # TODO: add -lmpi_gtl_cuda to avoid hangs (alps)
+            # '$PE_MPICH_GTL_DIR_nvidia80 $PE_MPICH_GTL_LIBS_nvidia80'
+        )
+        self.build_system.options = [
+            f'CUDA_INSTALL_PATH={cuda_path}',
+            'MPI_HOME=$CRAY_MPICH_PREFIX',
```
Comment (teojgo) [inline checks/prgenv/cuda/cuda_aware_mpi.py:222]:
```suggestion
            f'MPI_HOME=$CRAY_MPICH_PREFIX',
```

File: checks/prgenv/cuda/cuda_aware_mpi.py [hunk 9]
```diff
@@ -138,3 +139,129 @@ def set_num_tasks(self):
         self.num_tasks_per_node = 1
         self.num_gpus_per_node = 1
         self.executable_opts = ['-t %d 1' % self.num_tasks]
```
Comment (teojgo) [inline checks/prgenv/cuda/cuda_aware_mpi.py:141]:
```suggestion
        self.executable_opts = [f'-t {self.num_tasks} 1']
```

File: checks/prgenv/cuda/cuda_aware_mpi.py [hunk 10]
```diff
@@ -138,3 +139,129 @@ def set_num_tasks(self):
...
+        # nvhpc/22.3 has cuda/11.6 which supports gcc<12:
+        nvhpc2gcc = {
+            '21.3': {'cuda': '11.2', 'gcc': '10'},
+            '21.5': {'cuda': '11.3', 'gcc': '10'},
+            '21.9': {'cuda': '11.4', 'gcc': '11'},
+            '22.2': {'cuda': '11.6', 'gcc': '11'},
+            '22.3': {'cuda': '11.6', 'gcc': '11'},
+            # TODO: newer nvhpc
+        }
+        gcc_major_version = self.gcc_info.split("/")[1].split(".")[0]
```
Comment (teojgo) [inline checks/prgenv/cuda/cuda_aware_mpi.py:197]:
```suggestion
        gcc_major_version = self.gcc_info.split('/')[1].split('.')[0]
```

File: checks/prgenv/cuda/cuda_aware_mpi.py [hunk 11]
```diff
@@ -138,3 +139,129 @@ def set_num_tasks(self):
...
+            f'CUDA_INSTALL_PATH={cuda_path}',
+            'MPI_HOME=$CRAY_MPICH_PREFIX',
+            f'GENCODE_FLAGS="{gcd_flgs}"',
+            f'MPICC="{self.current_environ.cc}"',
+            f'MPILD="{self.current_environ.cxx} {link_flags}"',
+        ]
+
+    @run_before('compile')
+    def extract_versions(self):
+        cmd1 = "ldd ../bin/jacobi_cuda_aware_mpi | grep libcudart "
```
Comment (teojgo) [inline checks/prgenv/cuda/cuda_aware_mpi.py:230]:
```suggestion
        cmd1 = 'ldd ../bin/jacobi_cuda_aware_mpi | grep libcudart '
```

File: checks/prgenv/cuda/cuda_aware_mpi.py [hunk 12]
```diff
@@ -138,3 +139,129 @@ def set_num_tasks(self):
...
+        nvhpc2gcc = {
+            '21.3': {'cuda': '11.2', 'gcc': '10'},
+            '21.5': {'cuda': '11.3', 'gcc': '10'},
+            '21.9': {'cuda': '11.4', 'gcc': '11'},
+            '22.2': {'cuda': '11.6', 'gcc': '11'},
+            '22.3': {'cuda': '11.6', 'gcc': '11'},
+            # TODO: newer nvhpc
+        }
+        gcc_major_version = self.gcc_info.split("/")[1].split(".")[0]
+        nvhpc_version = self.nvhpc_info.split("/")[1]
```
Comment (teojgo) [inline checks/prgenv/cuda/cuda_aware_mpi.py:198]:
```suggestion
        nvhpc_version = self.nvhpc_info.split('/')[1]
```

File: checks/prgenv/cuda/cuda_aware_mpi.py [hunk 13]
```diff
@@ -138,3 +139,129 @@ def set_num_tasks(self):
...
+            f'cuda/{cuda_max_version}:gcc/{gcc_max_version}'
+        )
+        self.skip_if(gcc_major_version != gcc_max_version, skip_msg)
+        self.modules = [self.cdt_info, self.nvhpc_info, self.gcc_info]
+
+    @run_before('compile')
+    def set_compilers(self):
+        gput = self.current_partition.select_devices('gpu')[0]
+        gcd_flgs = f'-arch={gput.arch}'
+        nvhpc_version = self.nvhpc_info.split("/")[1]
```
Comment (teojgo) [inline checks/prgenv/cuda/cuda_aware_mpi.py:212]:
```suggestion
        nvhpc_version = self.nvhpc_info.split('/')[1]
```

File: checks/prgenv/cuda/cuda_aware_mpi.py [hunk 14]
```diff
@@ -138,3 +139,129 @@ def set_num_tasks(self):
...
+            f'GENCODE_FLAGS="{gcd_flgs}"',
+            f'MPICC="{self.current_environ.cc}"',
+            f'MPILD="{self.current_environ.cxx} {link_flags}"',
+        ]
+
+    @run_before('compile')
+    def extract_versions(self):
+        cmd1 = "ldd ../bin/jacobi_cuda_aware_mpi | grep libcudart "
+        cmd2 = "awk '{print \"ls -l \",$3}'"
+        cmd3 = "sh"
```
Comment (teojgo) [inline checks/prgenv/cuda/cuda_aware_mpi.py:232]:
```suggestion
        cmd3 = 'sh'
```

**Review by teojgo — CHANGES_REQUESTED** [review]

File: checks/prgenv/cuda/cuda_aware_mpi.py (hunk 7 above)
Comment (jgphpc) [reply checks/prgenv/cuda/cuda_aware_mpi.py:234]: I need self.rpt in report_linked_versions below

File: checks/prgenv/cuda/cuda_aware_mpi.py [hunk 15]
```diff
@@ -137,4 +138,127 @@ def set_num_tasks(self):
         self.num_tasks = 2
         self.num_tasks_per_node = 1
         self.num_gpus_per_node = 1
-        self.executable_opts = ['-t %d 1' % self.num_tasks]
+        self.executable_opts = [f'-t %d {self.num_tasks}']
```
Comment (teojgo) [inline checks/prgenv/cuda/cuda_aware_mpi.py:141]:
```suggestion
        self.executable_opts = [f'-t {self.num_tasks} 1']
```

**Review by teojgo — CHANGES_REQUESTED** [review]

File: checks/prgenv/cuda/cuda_aware_mpi.py (hunk 15 above)
Comment (jgphpc) [reply checks/prgenv/cuda/cuda_aware_mpi.py]: sorry!

File: checks/prgenv/cuda/cuda_aware_mpi.py [hunk 16]
```diff
@@ -137,4 +138,127 @@ def set_num_tasks(self):
         self.num_tasks = 2
         self.num_tasks_per_node = 1
         self.num_gpus_per_node = 1
-        self.executable_opts = ['-t %d 1' % self.num_tasks]
+        self.executable_opts = [f'-t %d {self.num_tasks} 1']
```
Comment (teojgo) [inline checks/prgenv/cuda/cuda_aware_mpi.py:141]:
```suggestion
        self.executable_opts = [f'-t {self.num_tasks} 1']
```

Remove the %d

**Review by teojgo — CHANGES_REQUESTED** [review]

**Review by teojgo — APPROVED** [review]

**Pull request merged**